### PR TITLE
Optimize product blocks

### PR DESCRIPTION
**Pull request by mnakalay**
I added thumbnail types for products primary and secondary images as well as product modal image
blocks and product modals are now using thumbnail types with fallback on legacy thumbnailer if necessary

I also cleaned up a bit the blocks, removed unused resources, redundancies...

### CODE REVIEW

**Comment by Mesuva** [comment]
Huge amount of work here, will review soon

**Comment by mnakalay** [comment]
if it can ake it easier for you to review you can Skype me and I'll walk you through it so you don't need to look at every line

**Comment by Mesuva** [comment]
Hit this with a 5.7 upgrade:

screen shot 2017-12-27 at 4 01 34 pm

**Comment by mnakalay** [comment]
oups. I'm on it

**Comment by mnakalay** [comment]
It should work now

**Comment by Mesuva** [comment]
That's working for me in 5.7.

All looking good so far. My only concern is with the thumbnails system overall (and not with these changes here).. I really like the thumbnail system for control over cropping, etc, but I'm wondering about the potential explosion of additional thumbnail files.  I guess if someone is running a shop the images in their file manager are primarily going to be for products, but that might not always be the case.  

I also find it a bit frustrating how it automatically created the retina versions of sizes.

Ultimately someone can still delete these extra thumbnail types if they don't want them to be generated and things will still work.

**Comment by Mesuva** [comment]
@MrKarlDilkington pointed out to me that the retina versions of thumbnails can be turned off now at /index.php/dashboard/system/files/thumbnails/options, so that's part of my concern gone

**Comment by Remo** [comment]
I'm still skeptical about thumbnail types, we've had a ton of issues with it. That's why I made those retina types optional too. If this was implemented https://github.com/concrete5/concrete5/issues/3559, I'd definitely support it, but we are already struggling with thumbnail types. It might be nice to be able to select your own thumbnail types, that way you could select an existing one instead of just adding more?

**Comment by mnakalay** [comment]
@Remo, conditional thumbnail types would be an absolute game changer and I agree as it is the thumbnailing system has issues, mostly due to the server space it requires.

Really the reason I modified CS to use thumbnail types is that someone was having problems and it was image related and seemed to be due to CS. Turns out it's most likely a C5 problem.

Having said so, being able to load a product list with already existing images instead of images resized on the fly seems like a good idea.

As for the thumbnail types used for this, I don't think it is necessary to allow users to select their own thumbnail types.

My understanding of Community Store is that it is a non-opinionated e-Commerce tool so it uses minimal styling and sensible defaults so the non-technical user can get by easily and the technical user can modify stuff easily.

Originally the 2 blocks included legacy thumbnailer code with hardcoded width and height values. I simply created thumbnail types to reflect those values.

But then I made sure the thumbnail type instances were at the top of each block's view so using your own thumbnail types is as easy as making a template and changing the 2 types handles for your own.

Last but not least, I see lots of developers not really aware of thumbnail types and how they work in the background. And from my experience, non-technical users have absolutely no clue it's even there so offering an option to temper with something you know nothing about might create even more problems.

**Comment by Mesuva** [comment]
I'm still really torn with this one, I think that if I was updating to this version on an existing site I'd want to remove the thumbnails immediately - not because I don't think it's a good feature, but because I wouldn't be using them unless a particular scenario came up where images needed to be cropped and zoomed a lot, and I wouldn't want the filesystem being filled up with lots of potentially unused version.

In other words, I feel that thumbnail use it something a bit more opt-in than opt out... hmmm..

**Comment by mnakalay** [comment]
so you'd want to make this dependent on a setting?

**Comment by Mesuva** [comment]
Potentially, I'm not quite sure.

I like the like the idea of having a setting in the shop where you can enable the thumbnail usage, where it installs the thumbnail types when you elect to.

Then the block templates could check for this setting before even looking for the thumbnails. 

(not saying I'm asking you to do this)

The big issue Im seeing is just this being enabled by default, with users/devs not even aware that it happened through an update, and then file usage going up dramatically.

**Comment by Mesuva** [comment]
I think this was held off until there was more changes in V8 that enabled better control over thumbnail generation. I believe that this is in place now, so it would be worth re-evaluating this.

As a sidenote, there's quite a number of changes coming via a development branch here:
https://github.com/concrete5-community-store/community_store/tree/development

This is a combination of a few features like decimal quantities and other minor features/fixes.
I'm away on holidays at the moment, but the plan is to merge this branch into master soon, before working on other new development items. It's really just about testing this branch.

**Comment by jasteele12** [comment]
Bumping this and also asking what is the best way to install the latest for testing and possible production with 8.4.3 (and hopefully the upcoming c5 release)?

**Comment by Mesuva** [comment]
I'd like to look again at this soon, just not of the highest priority since it's easy-ish to create custom templates.

Regarding the latest for testing and production, I'd be suggesting to use the V8 only branch and not the master at this point. It's now very stable, and they'll likely be a release soon for it.
(i'm using it for sites soon to be in production myself)

**Comment by Mesuva** [comment]
I doubt it, I think this probably should be closed for now. I left it open more as an indicator that it would be great to have, but I believe it would be need to be re-worked.

**Comment by JohntheFish** [comment]
Thanks. I also posted a qu+link to this mon slack with a question about product variant images.

**Comment by mnakalay** [comment]
If any of you want to let me know how and what aspect of it should be reworked, that could help...

As far as I remember, the thumbnail type part with fallback to legacy thumbnailer is pretty generic and should work.

The other stuff maybe not.

Any idea?
